src/main/java/de/tum/in/www1/artemis/web/rest/open/PublicProgrammingSubmissionResource.java

```diff
@@ -132,24 +132,25 @@ public ResponseEntity<Void> testCaseChanged(@PathVariable Long exerciseId, @Requ
         // as the VCS-server performs the request
         SecurityUtils.setAuthorizationObject();
 
-        String lastCommitHash = null;
-        try {
-            Commit commit = versionControlService.orElseThrow().getLastCommitDetails(requestBody);
-            lastCommitHash = commit.getCommitHash();
-            log.info("create new programmingSubmission with commitHash: {} for exercise {}", lastCommitHash, exerciseId);
-        }
-        catch (Exception ex) {
-            log.debug(
-                    "Commit hash could not be parsed from test repository from exercise {}, the submission will be created with the latest commit hash of the solution repository.",
-                    exerciseId, ex);
-        }
-
         // When the tests were changed, the solution repository will be built. We therefore create a submission for the solution participation.
-        ProgrammingSubmission submission = programmingSubmissionService.createSolutionParticipationSubmissionWithTypeTest(exerciseId, lastCommitHash);
+        SolutionProgrammingExerciseParticipation solutionParticipation = solutionProgrammingExerciseParticipationRepository.findByProgrammingExerciseIdElseThrow(exerciseId);
+        ProgrammingSubmission submission = programmingSubmissionService.getOrCreateSubmissionWithLastCommitHashForParticipation(solutionParticipation, SubmissionType.TEST);
+
         programmingMessagingService.notifyUserAboutSubmission(submission);
         // It is possible that there is now a new test case or an old one has been removed. We use this flag to inform the instructor about outdated student results.
         programmingTriggerService.setTestCasesChanged(exerciseId, true);
 
+        // Trigger rebuild of template submission. It cannot be done in the build plan since it is copied for every student submission
```

Comment:
Triggering the solution and template builds at the same time doesn’t sound right to me.

This endpoint is called by a webhook in the `tests`-repository, right? Then so far the (intended) process for Artemis was:

1. trigger solution build
2. receive result for solution build
3. update set of test cases for the exercises based on the ones existing for the solution
4. trigger a build for the template, so that the template is checked against this new set of test cases.

What could happen now:

1. trigger solution build
2. trigger template build
3. template build runs faster: result received for template build
4. create feedback for template build according to old test cases known to Artemis
5. result received for solution build
6. test cases are updated in Artemis ⇒ the template result no longer reflects the wanted state

Reply:
I agree. This could be a problem.